contributors/design-proposals/storage/online-grow-volume-size.md

```diff
@@ -0,0 +1,250 @@
+# Online Growing Persistent Volume Size
+
+**Authors:** [mlmhl](https://github.com/mlmhl)
+
+- [Overview](#overview)
+- [Goals](#goals)
+- [Non Goals](#non-goals)
+- [Use Cases](#use-cases)
+- [Prerequisite](#prerequisite)
+- [Implementation Design](#implementation-design)
+	- [Design Overview](#design-overview)
+	- [PersistentVolumeClaim Cache in Kubelet](#persistentvolumeclaim-cache-in-kubelet)
+	- [Online File System Resize in Kbuelet](#online-file-system-resize-in-kbuelet)
+		- [Volume Resize Request Detect](#volume-resize-request-detect)
+		- [File System Resizing Operation](#file-system-resizing-operation)
+
+## Overview
+
+Release 1.10 only supports offline file system resizing for PVCs, as this operation is only executed inside the `MountVolume` operation in kubelet. If a resizing request was submitted after the volume mounted, it won't be performed. This proposal intent to support online file system resizing for PVCs in kubelet.
+
+## Goals
+
+Enable users to increase size of PVC which already in use(mounted). The user will update PVC for requesting a new size.  Underneath we expect that - according kubelet will resize the file system for the PVC.
+
+## Non Goals
+
+* Offline file system resizing not included. If we found a volume need file system resizing but not mounted to the node yet, we will do nothing. This situation will be dealt with by the existing [offline file system resizing handler](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/grow-volume-size.md).
+
+* Extending resize tools: We only support for most common file systems' offline resizing in current release, and we prefer to stay the same for the online one.
+
+## Use Cases
+
+* As a user I am running Mysql on a 100GB volume - but I am running out of space, I should be able to increase size of volume mysql is using without losing all my data. (online and with data)
+* As a user I am running an application with a PVC. I should be able to resize the volume without losing data or mount point. (online and with data and without taking pod offline)
+
+## Prerequisite
+
+Currently we only support offline resizing for `xfs`, `ext3`, `ext4`. Online resizing of `ext3`,`ext4` is introduced from [Linux kernel-3.3](https://www.ibm.com/developerworks/library/l-33linuxkernel/), and `xfs` always supported growing mounted partitions (in-fact currently there is no way to expand an unmounted `xfs` file system), so they are all safe for online resizing. If user tries to expand a volume with other formats an error event will be reported for according pod.
```

Comment:
@saad-ali  @gnufied  I've added a section in here to illustrate the safety for online resizing, PTAL.

Reply:
Should this behavior be gated on a check of kernel version? If kernel < 3.3 then don't do online resizing?

Reply:
Yeah, it's better to add this check. Currently offline resizing is also performed after volume mounted as there is no way to expand an unmounted `xfs` file system, so I guess we can add this check inside [resizeFileSystem](https://github.com/kubernetes/kubernetes/blob/7243ac9091775eeb88dc32b6c97ea0aa4846354c/pkg/volume/util/operationexecutor/operation_generator.go#L590), then both online and offline operations can be gated. cc @gnufied

Reply:
I am not sure about adding this check. Linux Kernel 3.3 is really old.  Docker requirement is 3.10+ - https://docs.docker.com/install/linux/docker-ce/binaries/